pallets/stable-staking/src/lib.rs

```diff
@@ -0,0 +1,888 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		tokens::{
+			fungible::{Inspect as FInspect, Mutate as FMutate},
+			fungibles::{Inspect as FsInspect, Mutate as FsMutate},
+			Preservation,
+		},
+		StorageVersion,
+	},
+	PalletId,
+};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_runtime::{
+	traits::{AccountIdConversion, AtLeast32BitUnsigned, CheckedAdd, CheckedDiv, CheckedMul, One},
+	ArithmeticError, Perquintill, Saturating,
+};
+use sp_std::{collections::vec_deque::VecDeque, fmt::Debug, prelude::*};
+
+#[derive(PartialEq, Eq, Clone, Encode, Debug, Decode, TypeInfo)]
+pub struct StakingInfo<BlockNumber, Balance> {
+	// For a single position or
+	// Synetic overall average effective_time weighted by staked amount
+	effective_time: BlockNumber,
+	// Staked amount
+	amount: Balance,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Debug, Decode, TypeInfo)]
+pub struct StakingInfoWithOwner<AccountId, PoolId, StakingInfo> {
+	// For a single position or
+	// Synetic overall average effective_time weighted by staked amount
+	who: AccountId,
+	// Staked amount
+	pool_id: PoolId,
+	staking_info: StakingInfo,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Debug, Decode, TypeInfo)]
+pub struct StableRewardInfo<Balance> {
+	// Epoch index
+	epoch: u128,
+	// Staked amount
+	reward_amount: Balance,
+}
+
+impl<BlockNumber, Balance> StakingInfo<BlockNumber, Balance>
+where
+	Balance: AtLeast32BitUnsigned + Copy,
+	BlockNumber: AtLeast32BitUnsigned + Copy,
+{
+	// Mixing a new added staking position, replace the checkpoint with Synetic new one
+	// Notice: The logic will be wrong if weight calculated time is before any single added
+	// effective_time
+	fn add(&mut self, effective_time: BlockNumber, amount: Balance) -> Option<()> {
+		// We try force all types into u128, then convert it back
+		let e: u128 = effective_time.try_into().ok()?;
+		let s: u128 = amount.try_into().ok()?;
+
+		let oe: u128 = self.effective_time.try_into().ok()?;
+		let os: u128 = self.amount.try_into().ok()?;
+
+		let new_amount: u128 = os.checked_add(s)?;
+		// (oe * os + e * s) / (os + s)
+		let new_effective_time: u128 =
+			(oe.checked_mul(os)?.checked_add(e.checked_mul(s)?)?).checked_div(new_amount)?;
+		self.amount = new_amount.try_into().ok()?;
```

Comment:
why this fomular? I guess I'm very confused about what `effective_time` and `amount` exactly are 😂

Reply:
Recent added test should give enough sample of calculation process. Let's a go-through meeting in the next call.